transhumandesign/kag-base
# [added] Loser team's Tent explodes

## Status

- **READY**: this PR is (to the best of your knowledge) ready to be incorporated into the game.

## Description
```
[added] Loser team's tent explodes
```
Fixes https://github.com/transhumandesign/kag-base/issues/1712

This PR changes Tent to make it explode  as soon as there is a winner team and if the winner team doesn't equal the tent's team num. In other words, the loser team's tent explodes.
This also happens after a successful surrender vote.

I added more tent sprites to give it a fuse and a red flashing frame, like bomb.
I changed a few lines here and there to make it all work without changing existing logic.

The thing is that bombs will run `SetupBomb()` in their `onInit()` whereas tent will run it in `onTick()` (at a later point).
A check `if (this.hasTag("exploding"))` in `onInit()` in `BombTimer.as` was added because otherwise `BombFuseOn()` would be executed for the tents as soon as the map loads in online. 

Some instances of `getNet().isServer()` changed to `isServer()`.

Since some sprite stuff is done in the tent .as file, I renamed it from `TentLogic.as` to `Tent.as`.

I made it so the "no build" area at the tent's location is removed as soon as "shes_gonna_blow" animation plays, so the ground can take damage and other stuff can be built and interacted with there after tent is gone. Other "no build" areas still remain active.

Tent's explosion is basicly a bomb explosion with higher damage and slightly larger radius. Anyone who is very close will die and gib, and anyone who is somewhat close will die without gibbing.

I encountered a problem that in ctf tutorial when the map reloads, the game would assume the tents to have an s32 `bomb_timer` value of 0 for some reason so `onTick(CSprite@ this)` fires and changes the tents sprite frames unwantedly.
So I had to set `this.set_s32("bomb_timer", -1);` in `onInit(CBlob@ this)` in `Tent.as` as a counter-measure.

Added support for an offset for where the fuse sparks should come out. 
`this.set_Vec2f("custom_sparks_offset", Vec2f(0,-24));` is used in Tent's `onInit(CBlob@ this)` which makes the sparks come out at the top.
If no offset is given the sparks will come out in the blob's center position as before.

I refrained from implementing some ideas I had here - https://github.com/transhumandesign/kag-base/issues/1712#issuecomment-1765439626

Tested in offline and online, no known remaining bugs. I suggest to test it with other players before adding it to a live build.

## Idea for improvement

Although not necessary for now it's of course possible to reduce on the sprite frames and make the fuse as a spritelayer. But I'm not sure if it can be easily done without messing with the existing logic too much.

## Steps to Test or Reproduce

Play CTF or CTF tutorial online or offline.
Collect the other team's flags or make it so a surrender vote gets passed.
Notice the loser team's tent will explode and deal damage.
